Source/Session.swift

```diff
@@ -0,0 +1,497 @@
+                                              encodingMemoryThreshold: encodingMemoryThreshold,
+                                              request: request,
+                                              fileManager: fileManager,
+                                              multipartBuilder: multipartFormData)
+
+        return upload(multipartUpload)
+    }
+
+    // MARK: - Internal API
+
+    // MARK: Uploadable
+
+    func upload(_ uploadable: UploadRequest.Uploadable, with convertible: URLRequestConvertible) -> UploadRequest {
+        let uploadable = Upload(request: convertible, uploadable: uploadable)
+
+        return upload(uploadable)
+    }
+
+    func upload(_ upload: UploadConvertible) -> UploadRequest {
+        let request = UploadRequest(convertible: upload,
+                                    underlyingQueue: rootQueue,
+                                    serializationQueue: serializationQueue,
+                                    eventMonitor: eventMonitor,
+                                    delegate: self)
+
+        perform(request)
+
+        return request
+    }
+
+    // MARK: Perform
+
+    func perform(_ request: Request) {
+        switch request {
+        case let r as DataRequest: perform(r)
+        case let r as UploadRequest: perform(r)
+        case let r as DownloadRequest: perform(r)
+        default: fatalError("Attempted to perform unsupported Request subclass: \(type(of: request))")
+        }
+    }
+
+    func perform(_ request: DataRequest) {
+        requestQueue.async {
+            guard !request.isCancelled else { return }
+
+            self.performSetupOperations(for: request, convertible: request.convertible)
+        }
+    }
+
+    func perform(_ request: UploadRequest) {
+        requestQueue.async {
+            guard !request.isCancelled else { return }
+
+            do {
+                let uploadable = try request.upload.createUploadable()
+                self.rootQueue.async { request.didCreateUploadable(uploadable) }
+
+                self.performSetupOperations(for: request, convertible: request.convertible)
+            } catch {
+                self.rootQueue.async { request.didFailToCreateUploadable(with: error) }
+            }
+        }
+    }
+
+    func perform(_ request: DownloadRequest) {
+        requestQueue.async {
+            switch request.downloadable {
+            case let .request(convertible):
+                self.performSetupOperations(for: request, convertible: convertible)
+            case let .resumeData(resumeData):
+                self.rootQueue.async { self.didReceiveResumeData(resumeData, for: request) }
+            }
+        }
+    }
+
+    func performSetupOperations(for request: Request, convertible: URLRequestConvertible) {
+        do {
+            let initialRequest = try convertible.asURLRequest()
+            rootQueue.async { request.didCreateURLRequest(initialRequest) }
```

Comment:
As the implementation of this method uses async, the `Session.default.request(urlRequest).request` object doesn't have the same properties as of `urlRequest` when accessed immediately.

I have a broken test at https://github.com/Restofire/Restofire/blob/2abd788ea9e5454f5f5bf90ce1ae386028b2c4b3/Tests/AlamofireRequestable/ARequestableSpec.swift#L41

Reply:
Yes. If you need to test such properties, you can either make a request and check the properties when it's complete, or use an `EventMonitor` to tap into the request creation events.

Reply:
Thanks for quick response.

Just thinking, doesn't it make sense at the time of creation of `Request` object, it should give the same `urlRequest` object back immediately but if there is a side-effect then it should listen using the `EventMonitor`.?

Right now the following code will return false
```swift
let urlRequest = URLRequest(...)
let header = HTTPHeaders.authorization(username: "user", password: "password")
header.forEach { urlRequest.setValue($0.value, forHTTPHeaderField: $0.key) }
print(Session.default.request(urlRequest).request.allHTTPHeaderFields == urlRequest.allHTTPHeaderFields)
```

Reply:
Is it really unexpected when you realize the request creating process is now asynchronous? I suppose we could update the inline docs to indicate that it won't be available until request creation process is complete, and even then, that property is actually just the last of all `URLRequest`s received by the `Request`, so there's more complexity there than before, where intermediate `URLRequests` were lost. Additionally, testing that the Alamofire property has some value doesn't actually guarantee that the underlying `URLSession` used all of those properties to make a request. Certain headers, especially authorization headers, are specifically documented as not necessarily being used from the `URLRequest` but instead can be changed by the session. In any event, if you really need this test, I think waiting for the event is the right way to go, as it largely replicate the previous behavior.

Reply:
I realize the request creating process is now asynchronous. The urlRequest is something which is already created and I am passing to Request and expecting it to be the same.

The following code returns true
```swift
print(try! Session.default.request(urlRequest).convertible.asURLRequest().allHTTPHeaderFields == urlRequest.allHTTPHeaderFields)
```

Reply:
FYI - Immediately calling `print(request.debugDescription)` also broke as it uses the last request property too.

Reply:
Yes, we know. It should print "$ curl command could not be created" until everything is available.

Reply:
👍🏻